File: android/HelloWorld/EditAndNormalize/src/main/java/com/dynamsoft/ddn/editandnormalize/fragments/ScannerFragment.java
package com.dynamsoft.ddn.editandnormalize.fragments;

import android.app.AlertDialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.dynamsoft.core.basic_structures.CompletionListener;
import com.dynamsoft.core.basic_structures.Quadrilateral;
import com.dynamsoft.cvr.CaptureVisionRouter;
import com.dynamsoft.cvr.CaptureVisionRouterException;
import com.dynamsoft.cvr.CapturedResultReceiver;
import com.dynamsoft.cvr.EnumPresetTemplate;
import com.dynamsoft.dce.CameraEnhancer;
import com.dynamsoft.dce.CameraEnhancerException;
import com.dynamsoft.dce.EnumEnhancerFeatures;
import com.dynamsoft.ddn.DetectedQuadsResult;
import com.dynamsoft.ddn.editandnormalize.MainViewModel;
import com.dynamsoft.ddn.editandnormalize.R;
import com.dynamsoft.ddn.editandnormalize.databinding.FragmentScannerBinding;

import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.fragment.NavHostFragment;

public class ScannerFragment extends Fragment {

    private FragmentScannerBinding binding;
    private MainViewModel mViewModel;
    private CaptureVisionRouter mRouter;
    private CameraEnhancer mCamera;
    private boolean ifBtnClick;
    private AlertDialog mAlertDialog;


    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        binding = FragmentScannerBinding.inflate(inflater, container, false);
        mViewModel = new ViewModelProvider(requireActivity()).get(MainViewModel.class);
        initCaptureVisionRouter();
        return binding.getRoot();

    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding.btnCapture.setOnClickListener(v -> ifBtnClick = true);
        mCamera = new CameraEnhancer(binding.cameraView, getViewLifecycleOwner());
        try {
            // Enable the frame filter feature of Dynamsoft Camera Enhancer to process the high quality video frames only.
            mCamera.enableEnhancedFeatures(EnumEnhancerFeatures.EF_FRAME_FILTER);
            // You can use the following code to get normalized images with higher resolution.
            // mCamera.setResolution(EnumResolution.RESOLUTION_4K);
        } catch (CameraEnhancerException e) {
            e.printStackTrace();
        }
        try {
            // Set Dynamsoft Camera Enhancer as the input.
            mRouter.setInput(mCamera);
        } catch (CaptureVisionRouterException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        try {
            mCamera.open();
        } catch (CameraEnhancerException e) {
            e.printStackTrace();
        }
        mRouter.startCapturing(EnumPresetTemplate.PT_DETECT_DOCUMENT_BOUNDARIES, new CompletionListener() {
            @Override
            public void onSuccess() {

            }

            @Override
            public void onFailure(int errorCode, String errorString) {
                if(getActivity() != null){
                    getActivity().runOnUiThread(() -> showDialog("Error",
                            String.format(Locale.getDefault(),
                                    "ErrorCode: %d %nErrorMessage: %s", errorCode, errorString)));
                }
            }
        });
    }

    @Override
    public void onPause() {
        super.onPause();
        try {
            mCamera.close();
        } catch (CameraEnhancerException e) {
            e.printStackTrace();
        }
        mRouter.stopCapturing();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    private void showDialog(String title, String message) {
        if (mAlertDialog == null) {
            // Restart the capture when the dialog is closed
            mAlertDialog = new AlertDialog.Builder(getContext()).setCancelable(true).setPositiveButton("OK", null)
                    .create();
        }
        mAlertDialog.setTitle(title);
        mAlertDialog.setMessage(message);
        mAlertDialog.show();
    }

    private void initCaptureVisionRouter() {
        mRouter = new CaptureVisionRouter(requireActivity());
        // Add result receiver to receive callback when the result output.
        mRouter.addResultReceiver(new CapturedResultReceiver() {
            @Override
            public void onDetectedQuadsReceived(DetectedQuadsResult result) {
                if (ifBtnClick && result.getItems().length > 0) {
                    ifBtnClick = false;
                    mViewModel.capturedQuads = new Quadrilateral[result.getItems().length];
                    for (int i = 0; i < result.getItems().length; i++) {
                        mViewModel.capturedQuads[i] = result.getItems()[i].getLocation();
                    }
                    mViewModel.capturedWholeImage = mRouter.getIntermediateResultManager().getOriginalImage(result.getOriginalImageHashId());
                    mRouter.stopCapturing();
                    requireActivity().runOnUiThread(() -> NavHostFragment.findNavController(ScannerFragment.this)
                            .navigate(R.id.action_ScannerFragment_to_EditorFragment));
                }
            }
        });
    }

}

File: android/HelloWorld/EditAndNormalizeKt/src/main/java/com/dynamsoft/editandnormalizekt/fragments/ScannerFragment.kt
package com.dynamsoft.editandnormalizekt.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.NavHostFragment
import com.dynamsoft.core.basic_structures.CompletionListener
import com.dynamsoft.cvr.CaptureVisionRouter
import com.dynamsoft.cvr.CaptureVisionRouterException
import com.dynamsoft.cvr.CapturedResultReceiver
import com.dynamsoft.cvr.EnumPresetTemplate
import com.dynamsoft.dce.CameraEnhancer
import com.dynamsoft.dce.CameraEnhancerException
import com.dynamsoft.dce.EnumEnhancerFeatures
import com.dynamsoft.ddn.DetectedQuadsResult
import com.dynamsoft.editandnormalizekt.MainViewModel
import com.dynamsoft.editandnormalizekt.R
import com.dynamsoft.editandnormalizekt.databinding.FragmentScannerBinding

class ScannerFragment : Fragment() {
    private var _binding: FragmentScannerBinding? = null
    private val binding get() = _binding!!
    private lateinit var viewModel: MainViewModel
    private lateinit var router: CaptureVisionRouter
    private lateinit var camera: CameraEnhancer
    private var ifBtnClick = false
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _binding = FragmentScannerBinding.inflate(inflater, container, false)
        viewModel = ViewModelProvider(requireActivity())[MainViewModel::class.java]
        initCaptureVisionRouter()
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.btnCapture.setOnClickListener { v -> ifBtnClick = true }
        camera = CameraEnhancer(binding.cameraView, viewLifecycleOwner)
        try {
            // Enable the frame filter feature of Dynamsoft Camera Enhancer to process the high quality video frames only.
            camera.enableEnhancedFeatures(EnumEnhancerFeatures.EF_FRAME_FILTER)
            // You can use the following code to get normalized images with higher resolution.
            //camera.setResolution(EnumResolution.RESOLUTION_4K)
        } catch (e: CameraEnhancerException) {
            e.printStackTrace()
        }
        try {
            // Set Dynamsoft Camera Enhancer as the input.
            router.input = camera
        } catch (e: CaptureVisionRouterException) {
            e.printStackTrace()
        }
    }

    override fun onResume() {
        super.onResume()
        try {
            camera.open()
        } catch (e: CameraEnhancerException) {
            e.printStackTrace()
        }
        router.startCapturing(
            EnumPresetTemplate.PT_DETECT_DOCUMENT_BOUNDARIES,
            object : CompletionListener {
                override fun onSuccess() {}

                override fun onFailure(errorCode: Int, errorString: String?) {
                    activity?.runOnUiThread {
                        showDialog(
                            "Error",
                            "ErrorCode: $errorCode ErrorMessage: $errorString"
                        )
                    }
                }
            })
    }

    override fun onPause() {
        super.onPause()
        try {
            camera.close()
        } catch (e: CameraEnhancerException) {
            e.printStackTrace()
        }
        router.stopCapturing()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun showDialog(title: String?, vararg messages: String) {
        android.app.AlertDialog.Builder(context).setTitle(title)
            .setItems(messages, null)
            .setCancelable(true)
            .setPositiveButton("OK", null)
            .show()
    }

    private fun initCaptureVisionRouter() {
        router = CaptureVisionRouter(requireActivity())
        // Add result receiver to receive callback when the result output.
        router.addResultReceiver(object : CapturedResultReceiver {
            override fun onDetectedQuadsReceived(result: DetectedQuadsResult) {
                if (ifBtnClick && result.items.isNotEmpty()) {
                    ifBtnClick = false
                    viewModel.capturedQuads = arrayOfNulls(result.items.size)
                    for (i in result.items.indices) {
                        viewModel.capturedQuads!![i] = result.items[i].location
                    }
                    viewModel.capturedWholeImage = router.intermediateResultManager
                        .getOriginalImage(result.originalImageHashId)
                    router.stopCapturing()
                    requireActivity().runOnUiThread {
                        NavHostFragment.findNavController(this@ScannerFragment)
                            .navigate(R.id.action_ScannerFragment_to_EditorFragment)
                    }
                }
            }
        })
    }
}